library/core/src/alloc/layout.rs

```diff
@@ -172,27 +172,33 @@ impl Layout {
     /// allocate backing structure for `T` (which could be a trait
     /// or other unsized type like a slice).
     ///
+    /// This is an unchecked version of [`for_value`][Self::for_value]
+    /// which takes a raw pointer instead of a reference.
+    ///
     /// # Safety
     ///
-    /// This function is only safe to call if the following conditions hold:
-    ///
-    /// - If `T` is `Sized`, this function is always safe to call.
-    /// - If the unsized tail of `T` is:
-    ///     - a [slice], then the length of the slice tail must be an initialized
-    ///       integer, and the size of the *entire value*
-    ///       (dynamic tail length + statically sized prefix) must fit in `isize`.
-    ///     - a [trait object], then the vtable part of the pointer must point
-    ///       to a valid vtable for the type `T` acquired by an unsizing coercion,
-    ///       and the size of the *entire value*
-    ///       (dynamic tail length + statically sized prefix) must fit in `isize`.
-    ///     - an (unstable) [extern type], then this function is always safe to
-    ///       call, but may panic or otherwise return the wrong value, as the
-    ///       extern type's layout is not known. This is the same behavior as
-    ///       [`Layout::for_value`] on a reference to an extern type tail.
-    ///     - otherwise, it is conservatively not allowed to call this function.
+    /// The provided (possibly wide) pointer must describe a valid value layout.
+    /// Specifically:
+    ///
+    /// - If `T` is a `Sized` type, this function is always safe to call and is
+    ///   equivalent to [`size_of::<T>()`][size_of]. The pointer is unused.
+    /// - If the unsized tail of `T` is a [slice], then the size of the *entire
+    ///   value* (statically sized prefix plus dynamic tail) must fit in `isize`.
+    ///   The pointer may be dangling or null; only the slice length is used.
```

Comment:
"or unaligned or whatever" -- do we really want to list all that?

I'd prefer if we said something like "The data part of the pointer is unused", except we don't have standard terminology for "data part" AFAIK (I mean the part that's not the metadata).

Reply:
Saying something along the lines of "no read is performed" provides the property which is most directly relevant (and indirectly implies that the address is immaterial). So I'll suggest using

> The pointer does not need to be dereferenceable, as only the slice length is used.

In all honesty, I'd personally prefer exposing `unsafe fn(<T as Pointee>::Metadata) -> Layout` instead of `unsafe fn(*const T) -> Layout`, as it directly encodes this property, and I don't see read-performing `size_of` as particularly useful; both because of `*const UnsafeCell<ThinCStr>` not being able to provide it and because I've found explicit conversion between `&CStr` and `Thin<&CStr>` (a pointer wrapper stored as just `NonNull<()>`) to not cause issue in practice.

Reply:
Yeah that's fair. But then we won't have anything on stable until `Metadata` is stable...

Reply:
I changed it to

> The pointer does not need to be [valid](https://doc.rust-lang.org/stable/std/ptr/index.html#safety) for access, as only the pointer metadata is used.

This sticks to language already defined in the docs.